hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -5180,6 +5185,10 @@ private void replaceJdbcParametersType(
 						parameters.set( 0, newJdbcParameter );
 						jdbcParameters.getJdbcParameters().remove( jdbcParameter );
 						jdbcParameters.getJdbcParameters().add( newJdbcParameter );
+						for ( SqmParameterInterpretation interpretation : this.sqmInterpretationsBySqmParam.get( sqmParameter ) ) {
+							// We need to clear to make sure we overwrite eventual already resolved expressions
+							interpretation.clearResolvedExpression();
```

Comment:
I'm kind of surprised this is necessary. What is calling `SqmParameterInterpretation#getResolvedExpression()` so early that this must be reset? Maybe we should try to defer this call instead?

Reply:
For reference, `ArgumentTypesValidator.ParameterDetector` is what is initializing the resolved expression "early" - before we replace the jdbc paramter type.

Reply:
I updated the solution which now checks for `JavaObjectType` instead of walking the tree to detect jdbc parameters.